primitives/src/v2/mod.rs

```diff
@@ -1569,6 +1598,72 @@ impl CompactStatement {
 	}
 }
 
+/// `IndexedVec` struct indexed by type specific indices.
+#[derive(Clone, Encode, Decode, RuntimeDebug, TypeInfo)]
+#[cfg_attr(feature = "std", derive(PartialEq, MallocSizeOf))]
+pub struct IndexedVec<K, V>(Vec<V>, PhantomData<fn(K) -> K>);
+
+impl<K, V> Default for IndexedVec<K, V> {
+	fn default() -> Self {
+		Self(vec![], PhantomData)
+	}
+}
+
+impl<K, V> From<Vec<V>> for IndexedVec<K, V> {
+	fn from(validators: Vec<V>) -> Self {
+		Self(validators, PhantomData)
+	}
+}
+
+impl<K, V> FromIterator<V> for IndexedVec<K, V> {
+	fn from_iter<T: IntoIterator<Item = V>>(iter: T) -> Self {
+		Self(Vec::from_iter(iter), PhantomData)
+	}
+}
+
+impl<K, V> IndexedVec<K, V>
+where
+	V: Clone,
+{
+	/// Returns a reference to an element indexed using `K`.
+	pub fn get(&self, index: K) -> Option<&V>
```

Comment:
Why not take `K` by reference?

Reply:
No reason, just stuck with it. Can refactor if necessary.